hub/api/tensor.py

```diff
@@ -218,4 +221,14 @@ def __str__(self):
             index_str = ""
         return f"Tensor(key={repr(self.key)}{index_str})"
 
+    def __array__(self) -> np.ndarray:
+        if self._numpy is None:
+            self._numpy = self.numpy()
+            # NOTE: caching the array avoids multiple calls to .numpy() when a tensor object
```

Comment:
Inconsistency between the actual dataset value and the cached value here seems like a huge issue. I'd personally prefer no caching compared to this, subsequent .numpy() calls would anyway be faster due to the LRUCache.

Reply:
The cache gets invalidated on tensor modification.

Reply:
yeah, doesn't it make more sense from numpy's perspective that calling np.array(...) on something copies it?

Reply:
<img width="254" alt="Screen Shot 2021-07-12 at 3 26 57 PM" src="https://user-images.githubusercontent.com/26421036/125363527-a109c800-e325-11eb-989e-df7668854cc4.png">